### PR TITLE
TypeScript: remove extra indentation for arrow function on variable declaration with comment

### LINKED ISSUE #7066: [TS parser] Incorrect leading spaces inserted
**Prettier 1.19.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLAtgBwgJxgAgCp8BDAZ3zgA84wT8AzHCdfAcmtpLYB0oNsefAHMIjZqzYA6APSjeUPpChkC6Eqij4AvKTIBPKGHwAKAJQ6AfPmB989-MtWUax3aJOc6Zuw6cEwdAATHRcuX3snCAAbOCloiGETCIcSAHcNAi8TNkxNYQAvNgAafABtNgBaMBL2AFZatgAONgBdUtstB26SaOikfBgcAFc4YpT7AF8fLvsZqb5fGRl8ACVhqCg4HD4AbhNyQ2MGDbAYVGhzG18h-WvZ0gzUNQ0oc19JxxIYMAALUzgFk63Rgv2YaUoHz403MUjoP1+Jk8Fm01iisSk22YOGRMxAxRAEEw52gZGQoBIOHBAAVKQgySgSAA3CCoIL4kAAIxwJDAAGs4DAAMqYXn5ZBDUYE34wdDRADqv2ecDIorAcCF9OeqCZz30yHAZDJBM0ZG2MGpPOE6mQDF6ZoJACsyFQAEI8-mCoUkdBwAAymjgtvtYxAzqoQvysQAisMIPBg9EHSBRTgzTgDTB9JgVWAcKhiRzMPnYPK2aDkE0AAwE4sQM3ynmYA3FlXbJlBgkARzj8EtRIZIHIlS2cCCY45ODgPdQU8tJGtJETybN6FQy9DZCjcFj8aDSEloZgJE5ZaCFaQACYCUMNNF8gBhFg2lBQaCdkDDM0AFRPDLtSahkyowAJJQOOsBCnmBYwAAguBQpZrEG6TJMQA)
```sh
--parser typescript
--no-semi
--single-quote
```

**Input:**
```tsx
import * as execa from 'execa'
import go from './go'

const main = async () => {
    const exec = go(execa)
    const cmd = execa
    console.log(
      await exec('pingz', ['-c', '5', '8'], {
        all: true,
      })
    )
  }

  // Runner
;(async function() {
  try {
    await main()
  } catch (e) {
    throw e
  }
})().catch((e) => console.error(e))

```

**Output:**
```tsx
import * as execa from 'execa'
import go from './go'

const main = async () => {
    const exec = go(execa)
    const cmd = execa
    console.log(
      await exec('pingz', ['-c', '5', '8'], {
        all: true
      })
    )
  }

  // Runner
;(async function() {
  try {
    await main()
  } catch (e) {
    throw e
  }
})().catch(e => console.error(e))

```

**Expected behavior:**

Note that the main block is indented and so is the Runner comment. This only happens with the TypeScript parser.

### PR DESCRIPTION
**Pull request by sosukesuzuki**
Fix #7066 

```js
// Input
const foo = () => {
  return;
}

// foo
;

// Prettier stable
const foo = () => {
    return;
  };

  // foo

// Prettier master
const foo = () => {
  return;
};

// foo
```

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-js/printer-estree.js [hunk 1]
```diff
@@ -1726,7 +1726,7 @@ function printPathNoParens(path, options, print, args) {
       const hasValue = n.declarations.some(decl => decl.init);
 
       let firstVariable;
-      if (printed.length === 1 && !n.declarations[0].comments) {
+      if (printed.length === 1 && !hasLeadingComment(n.declarations[0])) {
```
Comment (sosukesuzuki) [inline src/language-js/printer-estree.js]: This condition is for code like below:
```js
const // comment
  foo = 3;
```
So, only target the leading comment.

Comment (thorn0) [reply src/language-js/printer-estree.js]: But why does it happen only with `typescript`?

Comment (sosukesuzuki) [reply src/language-js/printer-estree.js]: I found that the our comment attachment for babel's AST and ts-estree's AST is different in this case.

- For babel's AST: There is `n.comments`, there is not `n.declarations[0].comments`. (but, `n.declarations[0].trailingComments` does exists.)
- For ts-estree's AST: There is `n.declarations[0].comments`, there is not `n.comment`.

Probably this is a bug, I think we should fix comment attachment for Babel's AST.

Comment (thorn0) [reply src/language-js/printer-estree.js]: I investigated it a bit too. This difference is caused by the difference in location data for the `VariableDeclaration` node between the parsers.

**Prettier pr-7094**
[Playground link](https://deploy-preview-7094--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuc0DOMAEAzCFMC8mAFAJSEB8mwAOlJpgE5wwCujUA3HQL50D0-HHjrcoIADQgIABxgBLdMlABDRowgB3AApqEaZCBUA3CPIAmkkACNGKsAGsWAZRn35UAObIYjVnCkACxgAWwAbAHVA+Xg0NzA4Z30Y+WMYgE9DMDQDKQ80OEYYbTtPEJVkbBUwgqkAKzQADwAhO0cXFRC4ABkPOErq2pAGxucPTzC4AEVWCHgBmoCQN0YCxkMYdJk4NDBGeTkrGX3YCIsYQOQADgAGKWOIAoi7GUNjncLjfqkAR1n4EqyAwoFRoAC0UDgcHM0KszD+8mYJRUZQqSCqiykBRC8h8fiWaHGkxmc366MGSxgKmsZ3MF2QACYpL4VPIwuMAMIQELlQxQaDfECsAoAFWpwIxQ2M-gAklAYbBnHsDjAAILy5ybSYLAo8HhAA)
```sh
--parser typescript
```

**Input:**
```tsx
const foo = () => {
  return;
}
// foo
;

```

With `--parser typescript` it's 0...40. With `--parser babel` and `flow` it's 0...31.

Comment (thorn0) [reply src/language-js/printer-estree.js]: Turn's out there are already some hacks for this in `postprocess.js`. Must be very related.

Comment (thorn0) [reply src/language-js/printer-estree.js]: See `overrideLocEnd` there. Try replacing `options.parser === "flow"` with `Array.isArray(toBeOverriddenNode.range)` to make it work for TS too.

Comment (sosukesuzuki) [reply src/language-js/printer-estree.js]: Thanks, I see. done.

File: tests/typescript_arrow/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -40,3 +40,34 @@ app.get("/", (req, res): void => {
 
 ================================================================================
 `;
+
+exports[`comments.js 1`] = `
+====================================options=====================================
+parsers: ["typescript"]
+printWidth: 80
```
Comment (thorn0) [inline tests/typescript_arrow/__snapshots__/jsfmt.spec.js.snap]: Please add a test for `--no-semi`.

File: src/language-js/printer-estree.js (hunk 1 above)
Comment (thorn0) [reply src/language-js/printer-estree.js]: So this change in `printer-estree.js` is not needed now.

File: src/language-js/postprocess.js [hunk 3]
```diff
@@ -78,7 +78,7 @@ function postprocess(ast, options) {
   }
 
   function locEnd(node) {
-    return options.parser === "flow" ? node.range[1] : node.end;
+    return Array.isArray(node.range) ? node.range[1] : node.end;
```
Comment (thorn0) [inline src/language-js/postprocess.js]: Please remove this function and import `locEnd` from `./loc` instead.

**Comment by sosukesuzuki** [comment]
@prettier/core need more review

**Pull request merged**